### PR TITLE
Make wait() checks more reliable

### PR DESCRIPTION
**Pull request by Aparicio99**
**Reasoning** 

I've caught an issue where pre/post rotate scripts were being reported with an error "error running non-shared prerotate script for...", but only on very specific but deterministic conditions. After some debugging I've found it to be due to on my use case logrotate being launched from crond which had already inherited from its parent to ignore SIGCHLD.

Even in this situation, the issue depended on compiler optimization flags, and even on moving/deleting non related code around, which eventually found to be due to the uninitialized use of the wait() return argument, therefore the changes I propose for every use of wait()/waitpid().

---
**Commit message**

If wait()/waitpid() returns an error for some reason it will not write to the *wstatus argument. Checking wstatus in this case is non deterministic due to the variable not being initialized, and may lead to wrongly report the forked command as executed successfully.

Also, one possible reason for wait() to return an error even if the forked command has executed successfully is the child process exiting before we call wait(). Normally when the child process finishes it stays in a zombie state until the wait() call, but if our parent process has set to ignore the SIGCHLD signal, we will inherited it and our child processes will exit immediately. Restoring the signal handler on initialization will avoid this issue.

### CODE REVIEW

File: logrotate.c [hunk 1]
```diff
@@ -3292,6 +3292,11 @@ int main(int argc, const char **argv)
 
     message(MESS_DEBUG, "\nHandling %d logs\n", numLogs);
 
+    /* Restore SIGCHLD handler in case our parent process had it ignored.
+     * Not doing this may cause our forked childs to exit prematurely before our
+     * wait() call, making it impossible to known the child proper exit code */
+    signal(SIGCHLD, SIG_DFL);
```
Comment (cgzones) [inline logrotate.c:3298]: Please issue a warning on failure, e.g.:
```c
if (signal(SIGCHLD, SIG_DFL) == SIG_ERR)
    message(MESS_WARN, "failed to reset SIGCHLD handler: %s\n", strerror(errno));
```

**Review by cgzones — CHANGES_REQUESTED** [review]
Thanks, looks good to me except one small nit.

File: logrotate.c (hunk 1 above)
Comment (Aparicio99) [reply logrotate.c]: Updated with your suggestion.

File: logrotate.c [hunk 2]
```diff
@@ -545,8 +546,11 @@ static int runScript(const struct logInfo *log, const char *logfn, const char *l
         exit(1);
     }
 
-    wait(&rc);
-    return rc;
+    if (wait(&rc) > 0 && WIFEXITED(rc)) {
```
Comment (cgzones) [inline logrotate.c:549]: Could you please take this opportunity and replace with `wait(3)` with `waitpid(3)`.
```c
waitpid(pid, &rc, 0);
```

File: logrotate.c [hunk 3]
```diff
@@ -753,9 +757,7 @@ static int shred_file(int fd, const char *filename, const struct logInfo *log)
         exit(1);
     }
 
-    wait(&status);
-
-    if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+    if (wait(&status) < 0 || !WIFEXITED(status) || WEXITSTATUS(status)) {
```
Comment (cgzones) [inline logrotate.c:760]:
```c
waitpid(pid, &status, 0);
```

File: logrotate.c [hunk 4]
```diff
@@ -989,19 +991,19 @@ static int compressLogFile(const char *name, const struct logInfo *log, const st
     }
 
     close(compressPipe[0]);
-    wait(&status);
 
-    fsync(outFile);
-
-    if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+    if (wait(&status) < 0 || !WIFEXITED(status) || WEXITSTATUS(status)) {
```
Comment (cgzones) [inline logrotate.c:995]:
```c
waitpid(pid, &status, 0);
```

**Review by cgzones — COMMENTED** [review]
I'd like to take this opportunity and replace the `wait(3)` calls with `waitpid(3)`.
Also the `waitpid(3)` call on line 1112 is not checked.

**Comment by Aparicio99** [comment]
Replaced all wait() calls with waitpid() and added validation of all return codes.

strerror(errno) can also be added to the error messages, but since it would only apply to when waitpid() fails and not when "!WIFEXITED(status) || WEXITSTATUS(status)" is true, it would complicate the code and possible have 2 different error messages for each call. Not sure if it's worth it, but I could also add this if you think so.

**Comment by cgzones** [comment]
> Replaced all wait() calls with waitpid() and added validation of all return codes.

Thanks.

> strerror(errno) can also be added to the error messages, but since it would only apply to when waitpid() fails and not when "!WIFEXITED(status) || WEXITSTATUS(status)" is true, it would complicate the code and possible have 2 different error messages for each call. Not sure if it's worth it, but I could also add this if you think so.

What about a helper like:

```c
int waitpid_checked(pid_t pid, const char **errmsg) {
    int r, status;

    r = waitpid(pid, &status, 0);
    if (r <= 0) {
        *errmsg = strerror(errno);
        return -1;
    }

    if (!WIFEXITED(status)) {
        *errmsg = "child did not terminate normally";
        return -1;
    }

    if (WEXITSTATUS(status) != 0) {
        *errmsg = "child did not exit with status 0";
        return -1;
    }

    *errmsg = NULL;
    return 0;
}
```
?

**Comment by Aparicio99** [comment]
Implemented the helper as you suggested. Also updated runScript and callers to use the helper error message.

**Review by cgzones — APPROVED** [review]

**Comment by cgzones** [comment]
Applied. Thanks for your contribution and the patch iterations.

**Pull request merged**